### PR TITLE
MODLD-287

### PR DESCRIPTION
**Pull request by aleksei-pronichev**

### CODE REVIEW

File: src/main/java/org/folio/marc4ld/service/marc2ld/field/FieldControllerImpl.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+  }
+
+  private static boolean isMapping(Marc2ldFieldRule fieldRule, Marc2ldMapper mapper) {
+    return Optional.of(fieldRule)
+      .map(Marc2ldFieldRule::getPredicate)
+      .map(mapper::canMap)
+      .orElse(false);
+  }
+
+  public FieldMapper getMapper(DataField dataField, Collection<ControlField> controlFields, Marc2ldFieldRule rule) {
```
Comment (pkjacob) [inline src/main/java/org/folio/marc4ld/service/marc2ld/field/FieldControllerImpl.java:69]: private?

File: src/test/java/org/folio/marc4ld/mapper/test/Marc2Bibframe650IT.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+  @Test
+  void map_shouldContains_severalRegion() {
+    // given
+    var marc = loadResourceAsString("fields/marc_650.jsonl");
+
+    // when
+    var resource = marc2BibframeMapper.fromMarcJson(marc);
+
+    System.out.println(resource);
```
Comment (pkjacob) [inline src/test/java/org/folio/marc4ld/mapper/test/Marc2Bibframe650IT.java:33]: remove System.out

File: src/main/java/org/folio/marc4ld/service/marc2ld/field/property/merger/PropertyMergerFactoryImpl.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+  }
+
+  @Override
+  public PropertyMerger getConstant(Marc4BibframeRules.FieldRule rule) {
+    return constantPropertyMerger;
+  }
+
+  @Override
+  public PropertyMerger get(Marc4BibframeRules.FieldRule rule) {
+    if (isPlace(rule)) {
```
Comment (pkjacob) [inline src/main/java/org/folio/marc4ld/service/marc2ld/field/property/merger/PropertyMergerFactoryImpl.java:45]: Hi @aleksei-pronichev,

We should not have any special rules for "Place" resources. Is this done for creating multiple "subfocus" outgoing edges to "Place" resources as explained in MODLD-287?  Please note that the scenario explained in the ticket is just one example.

File: src/test/java/org/folio/marc4ld/mapper/test/Marc2Bibframe650IT.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+    assertThat(resource)
+      .isNotNull();
+
+    var subject = geSubject(resource);
+    assertThat(subject)
```
Comment (pkjacob) [inline src/test/java/org/folio/marc4ld/mapper/test/Marc2Bibframe650IT.java:39]: Can we also verify that two outgoing "subFocus" edges are created? One for "Italy" and the second for "Florence"

File: src/main/java/org/folio/marc4ld/service/marc2ld/field/property/transformer/PropertyTransformerFactoryImpl.java [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+@Component
+@RequiredArgsConstructor
+public class PropertyTransformerFactoryImpl implements PropertyTransformerFactory {
+
+  private final SeparatePropertyTransformer separatePropertyTransformer;
+  private final DefaultPropertyTransformer defaultPropertyTransformer;
+
+  @Override
+  public PropertyTransformer get(Marc4BibframeRules.FieldRule rule) {
+    if (rule.getTypes().contains("PLACE")) {
```
Comment (pkjacob) [inline src/main/java/org/folio/marc4ld/service/marc2ld/field/property/transformer/PropertyTransformerFactoryImpl.java:16]: Here also, no special property needed for "Place" alone.

File: src/main/java/org/folio/marc4ld/service/marc2ld/field/property/merger/PropertyMergerFactoryImpl.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+  public PropertyMergerFactoryImpl(
+    ConstantPropertyMerger constantPropertyMerger,
+    UnionMergeFunction unionMergeFunction,
+    SkipAdditionalMergeFunction skipAdditionalMergeFunction
+  ) {
+    this.constantPropertyMerger = constantPropertyMerger;
+    this.unionMergeFunction = unionMergeFunction;
+    this.skipAdditionalMergeFunction = skipAdditionalMergeFunction;
+    this.functions = new HashMap<>();
+    functions.put("GEOGRAPHIC_SUBDIVISION", unionMergeFunction);
```
Comment (pkjacob) [inline src/main/java/org/folio/marc4ld/service/marc2ld/field/property/merger/PropertyMergerFactoryImpl.java:34]: Here also, no special functions need to be applied for "GEOGRAPHIC_SUBDIVISION" and "GEOGRAPHIC_AREA_CODE"

File: src/main/java/org/folio/marc4ld/service/marc2ld/field/property/merger/PropertyMergerFactoryImpl.java (hunk 3 above)
Comment (aleksei-pronichev) [reply src/main/java/org/folio/marc4ld/service/marc2ld/field/property/merger/PropertyMergerFactoryImpl.java:45]: discussed

File: src/test/java/org/folio/marc4ld/mapper/test/Marc2Bibframe650IT.java (hunk 4 above)
Comment (aleksei-pronichev) [reply src/test/java/org/folio/marc4ld/mapper/test/Marc2Bibframe650IT.java:39]: a specially create 2 separate test
This is for control properties
next (in this file) for edges `map_shouldReturn_severalGeographicalNodes`

File: src/main/java/org/folio/marc4ld/service/marc2ld/field/property/transformer/PropertyTransformerFactoryImpl.java (hunk 5 above)
Comment (aleksei-pronichev) [reply src/main/java/org/folio/marc4ld/service/marc2ld/field/property/transformer/PropertyTransformerFactoryImpl.java:16]: Discussed, isolate in factory to understand all the cases

File: src/main/java/org/folio/marc4ld/service/marc2ld/field/property/merger/PropertyMergerFactoryImpl.java (hunk 6 above)
Comment (aleksei-pronichev) [reply src/main/java/org/folio/marc4ld/service/marc2ld/field/property/merger/PropertyMergerFactoryImpl.java:34]: discussed

File: src/main/java/org/folio/marc4ld/service/marc2ld/Marc2ldFieldRule.java [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+import org.folio.ld.dictionary.model.Resource;
+import org.folio.marc4ld.configuration.property.Marc4BibframeRules;
+import org.marc4j.marc.ControlField;
+import org.marc4j.marc.DataField;
+
+public interface Marc2ldFieldRule {
+
+  Marc4BibframeRules.FieldRule getOriginal();
+
+  Resource computeParentIfAbsent(Resource resource);
```
Comment (PBobylev) [inline src/main/java/org/folio/marc4ld/service/marc2ld/Marc2ldFieldRule.java:18]: Why a "Rule" computes some 'Resource'?
"Rule" sound like a container with details how some process should go. 
But not do the processing, right?

File: src/main/java/org/folio/marc4ld/service/marc2ld/Marc2ldFieldRule.java [hunk 8]
```diff
@@ -0,0 +1,50 @@
+package org.folio.marc4ld.service.marc2ld;
```
Comment (PBobylev) [inline src/main/java/org/folio/marc4ld/service/marc2ld/Marc2ldFieldRule.java:1]: Why a "Rule" is located in a "service" package?

File: src/main/java/org/folio/marc4ld/service/marc2ld/Marc2ldFieldRule.java (hunk 7 above)
Comment (PBobylev) [reply src/main/java/org/folio/marc4ld/service/marc2ld/Marc2ldFieldRule.java:18]: This comment is applicable also for methods selectResourceFromEdges, getLabel, createProperties, mergeProperties

File: src/main/java/org/folio/marc4ld/service/marc2ld/Marc2ldFieldRuleImpl.java [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+import org.marc4j.marc.ControlField;
+import org.marc4j.marc.DataField;
+
+@Builder
+@AllArgsConstructor
+public class Marc2ldFieldRuleImpl implements Marc2ldFieldRule {
+
+  private final String label;
+  private final Relation relation;
+  @NonNull
```
Comment (PBobylev) [inline src/main/java/org/folio/marc4ld/service/marc2ld/Marc2ldFieldRuleImpl.java:30]: What's the point of this annotation on the field level? It's said:
"If put on a field, any generated method assigning a value to this field will also produce these null-checks."
and there is no generated methods, right?

File: src/main/java/org/folio/marc4ld/service/marc2ld/field/FieldControllerImpl.java [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+  @Override
+  public void handleField(
+    Resource parent,
+    DataField dataField,
+    List<ControlField> controlFields,
+    Marc2ldFieldRule fieldRule
+  ) {
+    var fieldMapper = getMapper(dataField, controlFields, fieldRule);
+    var mappedResources = fieldMapper.createResources(parent);
+
```
Comment (PBobylev) [inline src/main/java/org/folio/marc4ld/service/marc2ld/field/FieldControllerImpl.java:38]: what's the point of empty line? It doesn't improve readability, it decreases it :)
(general comment for all similar cases)

File: src/main/java/org/folio/marc4ld/service/marc2ld/field/FieldControllerImpl.java [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+    mappedResources
+      .forEach(resource -> additionalMapping(resource, fieldRule, dataField, controlFields));
```
Comment (PBobylev) [inline src/main/java/org/folio/marc4ld/service/marc2ld/field/FieldControllerImpl.java:40]: what's the point of a new line here?
(general comment for all similar cases)

File: src/main/java/org/folio/marc4ld/service/marc2ld/field/FieldControllerImpl.java [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+public class FieldControllerImpl implements FieldController {
+
+  private final RelationProvider relationProvider;
+  private final Marc2ldMapperController marc2ldMapperController;
+  private final FingerprintHashService hashService;
+  private final MapperHelper mapperHelper;
+
+  @Override
+  public void handleField(
+    Resource parent,
```
Comment (PBobylev) [inline src/main/java/org/folio/marc4ld/service/marc2ld/field/FieldControllerImpl.java:31]: what's the point of a new line here?
(general comment for all similar cases)

**Review by PBobylev — CHANGES_REQUESTED** [review]

**Comment by pkjacob** [comment]
Hi @aleksei-pronichev, I did testing of the PR and found that repeating some of the subfields makes the tests to fail. For example, take a look at this change I made in full_marc.json - https://github.com/FOLIO-EIS/lib-linked-data-marc4ld/pull/91. It caused the test to fail. Can you please take a look?

File: src/main/java/org/folio/marc4ld/service/marc2ld/Marc2ldFieldRule.java (hunk 7 above)
Comment (aleksei-pronichev) [reply src/main/java/org/folio/marc4ld/service/marc2ld/Marc2ldFieldRule.java]: discussed, renamed to applier

File: src/main/java/org/folio/marc4ld/service/marc2ld/Marc2ldFieldRuleImpl.java (hunk 9 above)
Comment (aleksei-pronichev) [reply src/main/java/org/folio/marc4ld/service/marc2ld/Marc2ldFieldRuleImpl.java]: Discussed
it's lombok annotation for lombok `Builder`

File: src/main/java/org/folio/marc4ld/service/marc2ld/field/FieldControllerImpl.java (hunk 10 above)
Comment (aleksei-pronichev) [reply src/main/java/org/folio/marc4ld/service/marc2ld/field/FieldControllerImpl.java:36]: discussed, new solution was founded

File: src/main/java/org/folio/marc4ld/service/marc2ld/field/FieldControllerImpl.java (hunk 11 above)
Comment (aleksei-pronichev) [reply src/main/java/org/folio/marc4ld/service/marc2ld/field/FieldControllerImpl.java]: discussed

File: src/main/java/org/folio/marc4ld/service/marc2ld/field/FieldControllerImpl.java (hunk 12 above)
Comment (aleksei-pronichev) [reply src/main/java/org/folio/marc4ld/service/marc2ld/field/FieldControllerImpl.java]: discussed

File: src/main/java/org/folio/marc4ld/service/marc2ld/field/property/transformer/SeparatePropertyTransformer.java [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.springframework.stereotype.Component;
+
+@Component
+public class SeparatePropertyTransformer implements PropertyTransformer {
+
+  @Override
+  public Collection<Map<String, List<String>>> apply(Map<String, List<String>> properties) {
```
Comment (pkjacob) [inline src/main/java/org/folio/marc4ld/service/marc2ld/field/property/transformer/SeparatePropertyTransformer.java:14]: Hi @aleksei-pronichev,

I think this logic will cause issues for [marc field 651](https://www.loc.gov/marc/bibliographic/bd651.html) which is mapped to a Place resource.

In 651, the place name ($a) is non repeatable. However, Miscellaneous Info ($g) is repeatable. In such case, we will throw an exception in line 24.

Comment (aleksei-pronichev) [reply src/main/java/org/folio/marc4ld/service/marc2ld/field/property/transformer/SeparatePropertyTransformer.java:14]: add some test cases for
-651 (conflict)
-650 Place
-650 Form

**Review by PBobylev — APPROVED** [review]
Target this PR to the "post-alpha' branch please

**Pull request merged**